Add dydx page. (#79)
skip-mev/docs: docs/slinky/integrations/0-dydx.md

---
description: dYdX
title: dYdX
sidebar_position: 0
---

## FAQs

### **Q: How do I get up and running?**

> **A:** Here is a quick (5-10 minutes) step by step breakdown. Please reach out to us at [skip.money/discord](http://skip.money/discord) if you have questions!

Reviewer: ```suggestion
**A:** Here is a quick (5-10 minutes) step-by-step breakdown. Please reach out to us at [skip.money/discord](http://skip.money/discord) if you have questions!
```

1. **Download the slinky binary.**

   The best way to get the slinky binary is in the GitHub releases page for Slinky.

   https://github.com/skip-mev/slinky/releases

>    The initial version required for dYdX is `0.3.1`

Reviewer 2: Is the plan for us to update this everytime a new slinky version is pushed to dydx (i.e 0.4.0)?

Author: For now I think it's fine--can add to release documentation.

   https://github.com/skip-mev/slinky/releases/tag/v0.3.1

   We also provide a container image at [ghcr.io/skip-mev/slinky-sidecar](http://ghcr.io/skip-mev/slinky-sidecar)

   This will include the Slinky binary, `slinky` , and a utility binary `slinky-config` as well as some pre-generated configuration files.

2. **Integrate the Slinky sidecar into your setup.**

   The configuration of your validator setup requires you to tackle a few problems which we’ll mention here.

   **_Configure the Slinky process_**

>    Slinky has 1 important config file for running with dYdX.

Reviewer: ```suggestion
   Slinky has 1 important config file for running with dYdX:
```

   The default values (excluding the dynamic updating process detailed below) in the `oracle.json` included in the release have been tested by the Skip team. We recommend using these and working with us to understand and adjust individual values that might optimize your operations.

   **Setup Dynamic Updating**

   Because the set of prices the chain wants to fetch change frequently, you must run Slinky with dynamic config updating. Using the `slinky-config` binary you can generate an `oracle.json` file which watches the chain and updates Slinky when on-chain data is changed.

   As an example, running the following would point your sidecar binary at the `dydxprotocold` node binary running on the same host at the default port of `1317`:

   ```bash
   slinky-config --chain dydx --node-http-url "localhost:1317"
   ```

   This command (with default localhost:1317) should produce the equivalent of the `oracle.json` file bundled in the `config/dydx` directory in the release. After running this command you should have produced valid `oracle.json` file and you can start your sidecar process.

> 

Reviewer 2: Maybe useful to include the relative path of the output file

   enabled = "true"

   # Oracle Address is the URL of the out of process oracle sidecar. This is used to
   # connect to the oracle sidecar when the application boots up. Note that the address
   # can be modified at any point, but will only take effect after the application is
   # restarted. This can be the address of an oracle container running on the same
   # machine or a remote machine.
   oracle_address = "localhost:8080"

   # Client Timeout is the time that the client is willing to wait for responses from
   # the oracle before timing out.
   client_timeout = "2s"

   # MetricsEnabled determines whether oracle metrics are enabled. Specifically
   # this enables intsrumentation of the oracle client and the interaction between
   # the oracle and the app.
   metrics_enabled = "false"

   # PrometheusServerAddress is the address of the prometheus server that metrics will be
   # exposed to.
   prometheus_server_address = ""
   ```

### **Q: How do I know if my validator is properly fetching prices and posting them to the chain?**

**A:** A full set of prometheus metrics are integrated into both the sidecar and the dYdX application binary.

> A comprehensive overview of the relevant metrics is hosted in the root [metrics.md](https://github.com/skip-mev/slinky/blob/main/metrics.md).

Reviewer: do we want this link from main or to be from a hash?

- [Sidecar Provider Metrics](https://github.com/skip-mev/slinky/blob/a6577aa779f1f98c47ec95d626e4af70949af7b1/providers/base/metrics/README.md)
- [API Provider Metrics](https://github.com/skip-mev/slinky/blob/a6577aa779f1f98c47ec95d626e4af70949af7b1/providers/base/api/metrics/README.md)
- [Websocket Provider Metrics](https://github.com/skip-mev/slinky/blob/a6577aa779f1f98c47ec95d626e4af70949af7b1/providers/base/websocket/metrics/README.md)

Additionally, the logs from your dYdX node binary will contain the following error if it is unable to connect to slinky to grab prices:

`Failed to run fetch prices for slinky daemon`

### **Q: How do I upgrade the sidecar binary?**

**A:** Upgrading the sidecar can be done out of band of the chain’s binary. If you have a load balancer, CNAME, etc., in front of your sidecar you can simply start up the new version and switch out which version traffic is being directed to during live chain validation.

If you are running the slinky sidecar in a container you can shut down the container, pull the updated container image and relaunch your container to update.

> If you are running the binary via systemd or other management tool you will need to stop the process and re-launch using the newly released binary.

Reviewer: ```suggestion
If you are running the binary via systemd or other management tool, you will need to stop the process and re-launch using the newly released binary.
```

:::note
We recommend you build some automation around config management either by pulling the latest `oracle.json` file directly from the release (if your dydx node is at localhost:1317) or via reconstructing the config using the `slinky-config` binary included in the release. The `oracle.json` file from previous releases will be compatible with future releases unless there is a major version bump, _however_, newly added price feeds may require updated information from your oracle config that was not present in the previous release which may cause breakage.

> :::

Reviewer 2: nit would explicitly say which oracle.json to use, i.e in `config/dydx`

The dydx node will still be able to participate in consensus without the sidecar, and will begin attaching prices to blocks once slinky is available. In the worst case, an upgrade in any of these manners will cause you to miss including vote extensions for a single block which should have no negative effects on you or the network.

### **Q: Which version of the sidecar binary should I be running?**

> **A:** We are currently in the process of designing solutions around this problem in order to give validators reliable signal for exactly the minimum version of sidecar they need to be running.

Reviewer: ```suggestion
**A:** We are currently in the process of designing solutions around giving validators reliable signals for exactly the minimum version of sidecar they need to be running.
```

In the short term, sidecar version bumps will be scheduled well in advance of any required upgrade and comms will be sent out in the relevant validator channels with the exact version required, and any changes which are relevant to the operator experience.

> In the future as we work with the chain and validators to build more automation around version bumps and reasoning about compatibility based on on-chain data, we plan to increase the cadence of updates in line with feature additions, as opposed to batching for larger releases on a longer timeline.

Reviewer: this sentence is too long
